### PR TITLE
Allow ? in domain of a wildcard url

### PR DESCRIPTION
**Pull request by hmstepanek**
Allow ?'s inside the domain of a url like:
https://example.com?foo=bar
https://example.???.

Previously the user would have had to add the query after the domain by inserting a / into the url. This is confusing for users who aren't url experts to have to modify the original url. Aka: https://example.com?foo=bar => https://example.com/?foo=bar.

This is a fix for https://github.com/hypothesis/product-backlog/issues/810.

### CODE REVIEW

**Comment by robertknight** [comment]
> Allow a * at the end of the domain of the url like:

Since we're not checking for a TLD in the domain, it looks like this also allows usage like `http://www.*` or `http://e*`? If so, that basically defeats the reason we added this restriction in the first place - to avoid queries which ES ends up expanding to the majority of URLs in the index.

Can you recap what the main user-facing problem was that we were trying to solve originally? Maybe there is an alternative approach, and it might be different for the API vs activity pages.

**Comment by hmstepanek** [comment]
urlparse doesn't support TLDs. The only way I can see to do this is to grab a list of domains from an externally maintained source and that is something we don't want to do. Therefore-I think I'm going to forgo supporting things like http://example.com* for now and just say *'s are not allowed at all in the domain but ? are allowed. -I'll update this in the original ticket and once ? support in the domain is merged review this with some end users and get feedback.

**Comment by dwhly** [comment]
@robertknight:

> Can you recap what the main user-facing problem was that we were trying to solve originally? Maybe there is an alternative approach, and it might be different for the API vs activity pages.

FWIW, I think the key user-facing problem is that `example.com/*` will not return the entire set of results for example.com, since it will miss (for instance) `example.com?querystring`.  (And most importantly, i presume that the large majority of users may not know these results are missing and will assume that the count of results represents the full set for the whole domain).  I agree that while this is an important use case, shipping what we have is a big step forward and shouldn't be delayed.  We can let our priorities be driven by user feedback.

**Comment by hmstepanek** [comment]
To answer your original question Rob: https://hypothes-is.slack.com/archives/C4K6M7P5E/p1537231057000100?thread_ts=1537229767.000100&cid=C4K6M7P5E

In summary: it's difficult to explain to users (and we don't want to need to explain to them) why they can't put the * at the end of the url and instead have to put a /* when the original url didn't have an ending /. I think however, the more common scenario would be there is a query at the end of the original url without the / (http://example?foo=bar). So from my perspective it's more important to support ? than *, as not supporting ? in the domain would require the user to modify the original url (http://example/?foo=bar) which is much more confusing than the * scenario.

**Comment by hmstepanek** [comment]
> FWIW, I think the key user-facing problem is that example.com/* will not return the entire set of results for example.com, since it will miss (for instance) example.com?querystring. 

@dwhly That is no longer true (It was when we first started developing this feature but it was fixed after getting feedback from the team). It doesn't miss results from example.com?querystring.

**Comment by dwhly** [comment]
> That is no longer true (It was when we first started developing this feature but it was fixed after getting feedback from the team). It doesn't miss results from example.com?querystring.

Thanks for clarifying!

>  it's difficult to explain to users (and we don't want to need to explain to them) why they can't put the * at the end of the url and instead have to put a /* when the original url didn't have an ending /

Thanks for clarifying!

File: h/search/query.py [hunk 1]
```diff
@@ -22,19 +22,25 @@ def wildcard_uri_is_valid(wildcard_uri):
...
-
-        # Remove all parts of the url except the scheme, netloc, and provide a substitute
-        # path value "p" so that uri's that only have a scheme and path are still valid.
-        uri_parts = (normalized_uri.scheme, normalized_uri.netloc, "p", "", "", "")
+        normalized_uri = urlparse.urlparse(wildcard_uri.replace("*", ""))
+        # If the normalized uri has no path component but has a query component.
+        if (not normalized_uri.path and
+                normalized_uri.netloc and
+                normalized_uri.query):
+            uri_parts = (normalized_uri.scheme, normalized_uri.netloc, "", normalized_uri.query, "", "")
```
Comment (robertknight) [inline h/search/query.py]: The urlparse result is a `(scheme, netloc, path, params, query, fragment)`, so you are putting the _query_ into the _params_ section and then undoing the swap with the "/;" => "?" change. Was that intended?

File: docs/_extra/api-reference/hypothesis.yaml [hunk 2]
```diff
@@ -353,8 +353,9 @@ paths:
             PDF fingerprint.
 
             `*` will match any character sequence (including an empty one), 
-            and a `?` will match any single character. Wildcards are only permitted 
-            within the path and query parts of the URI.
+            and a `?` will match any single character. `*`s are only permitted 
+            within the path and query parts of the URI and `?`s are only permitted
+            within the netloc, path, and query parts of the URI.
```
Comment (robertknight) [inline docs/_extra/api-reference/hypothesis.yaml]: Hmm, although precise, "netloc" is quite a Python-specific term I think (although I'm not sure that it is exclusively used in Python). Perhaps "domain" or "hostname" would be clearer in this context for readers who are not familiar with Python.

**Comment by robertknight** [comment]
> FWIW, I think the key user-facing problem is that example.com/* will not return the entire set of results for example.com, since it will miss (for instance) example.com?querystring.

Given the way we currently normalize URLs and the way wildcard search works, you're right about this specific case. 

I think that's actually an error in how we do URL normalization, since `https://example.com?foobar` actually becomes `https://example.com/?foobar` when the browser makes a request to the server. In other words, if you don't specify a path the browser implicitly converts it to "/". Now is not a great time to change that however.

**Comment by hmstepanek** [comment]
Yes urlunparse doesn't add the / after the domain unless there is a path component.

File: h/search/query.py [hunk 3]
```diff
@@ -22,19 +22,25 @@ def wildcard_uri_is_valid(wildcard_uri):
...
-        normalized_uri = urlparse.urlparse(wildcard_uri.replace("*", "").replace("?", ""))
-
-        # Remove all parts of the url except the scheme, netloc, and provide a substitute
-        # path value "p" so that uri's that only have a scheme and path are still valid.
-        uri_parts = (normalized_uri.scheme, normalized_uri.netloc, "p", "", "", "")
+        normalized_uri = urlparse.urlparse(wildcard_uri.replace("*", ""))
+        # If the normalized uri has no path component but has a query component.
+        if (not normalized_uri.path and
+                normalized_uri.netloc and
+                normalized_uri.query):
```
Comment (robertknight) [inline h/search/query.py]: Note that this path won't be taken if the URL contains a "?" but the query is empty. So "https://bbc.com?" fails. I can't think of a practical scenario where that is going to be a problem though.

File: h/search/query.py [hunk 4]
```diff
@@ -22,26 +22,20 @@ def wildcard_uri_is_valid(wildcard_uri):
...
-
-        # If a wildcard was in the scheme the uri may come back as "" (a falsey value).
-        if begining_of_uri and wildcard_uri.startswith(begining_of_uri):
-            return True
-    except ValueError:
-        pass
-    return False
+
+    normalized_uri = urlparse.urlparse(wildcard_uri.replace("?", ""))
+    if not normalized_uri.scheme or "*" in normalized_uri.scheme or "*" in normalized_uri.netloc:
```
Comment (robertknight) [inline h/search/query.py]: `urlparse` will never return a scheme containing a `*`. See https://github.com/python/cpython/blob/d4c76d960b8b286b75c933780416ace9cda682fd/Lib/urllib/parse.py#L75

File: tests/h/search/query_test.py [hunk 5]
```diff
@@ -582,21 +582,21 @@ def _get_search(self, search, pyramid_request, separate_keys):
 
 
 @pytest.mark.parametrize('wildcard_uri,expected', [
-    ("http?://bar.com", False),
     ("htt*://bar.com", False),
+    ("*http://bar.com", False),
+    ("http?://bar.com", False),
+    ("http://bar?com*", False),
```
[the remark this answers is not in the document]
Comment (robertknight) [reply tests/h/search/query_test.py]: Ah right - I forgot that `?` becomes a wildcard in this context.

**Pull request merged**